packages/components/src/button/index.js

```diff
@@ -41,13 +46,15 @@ export function Button( props, ref ) {
 
 	const tag = href !== undefined && ! disabled ? 'a' : 'button';
 	const tagProps = tag === 'a' ? { href, target } : { type: 'button', disabled };
+	const { isWithinToolbar, ref: finalRef } = useIsWithinToolbar( ref );
+	const allProps = { ...tagProps, ...additionalProps, className: classes, ref: finalRef };
 
-	return createElement( tag, {
-		...tagProps,
-		...additionalProps,
-		className: classes,
-		ref,
-	} );
+	// If it's within a toolbar, render with toolbar context instead.
+	if ( isWithinToolbar ) {
```

Comment:
Yes! But also because it's impossible to provide a good migration path without greatly improving `ToolbarButton`. Ideally, `ToolbarButton` would be flexible enough for all use cases. Even on Gutenberg we use a lot of different types of buttons in the Toolbar. [I even tried to migrate them all to `ToolbarButton`](https://github.com/WordPress/gutenberg/pull/17875/commits/5a03322508cacab6d02956c6567ddcfc9e8b24ad), but right not it's not possible, so I reverted it and put the code on `Button` instead.

Reply:
Have you considered using:
```js
element.closest( 'role="toolbar"' );
```
 instead?
I don't know if it is any better, but I used in the dropdown implementation to detect whether the focus is transferred to the modal dialog with this check
```js
document.activeElement.closest( '[role="dialog"]' )
```

Reply:
I'm using `.closest` inside `useIsWithinToolbar`.